group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorService.java

```diff
@@ -524,37 +495,39 @@ public CompletableFuture<ListGroupsResponseData> listGroups(
             );
         }
 
-        final CompletableFuture<ListGroupsResponseData> future = new CompletableFuture<>();
-        final List<ListGroupsResponseData.ListedGroup> results = new ArrayList<>();
         final Set<TopicPartition> existingPartitionSet = runtime.partitions();
-        final AtomicInteger cnt = new AtomicInteger(existingPartitionSet.size());
 
         if (existingPartitionSet.isEmpty()) {
             return CompletableFuture.completedFuture(new ListGroupsResponseData());
         }
 
+        final List<CompletableFuture<List<ListGroupsResponseData.ListedGroup>>> futures =
+            new ArrayList<>();
+
         for (TopicPartition tp : existingPartitionSet) {
-            runtime.scheduleReadOperation(
+            futures.add(runtime.scheduleReadOperation(
                 "list-groups",
                 tp,
                 (coordinator, lastCommittedOffset) -> coordinator.listGroups(request.statesFilter(), lastCommittedOffset)
-            ).handle((groups, exception) -> {
-                if (exception == null) {
-                    synchronized (results) {
-                        results.addAll(groups);
-                    }
+            ).exceptionally(exception -> {
+                exception = Errors.maybeUnwrapException(exception);
```

Comment:
did we not unwrap the exception correctly before 😓

Reply:
The handling was actually correct here. However, I discovered recently that the `exception` could be wrapped in some case so I got a bit paranoiac and unwrap everywhere.

Reply:
I see. Makes sense.